### ISSUE TITLE
K3s Prometheus unable to scrape worker nodes. 

### ISSUE THREAD

**Issue by james-millner** [carlosedp/cluster-monitoring#30]
Hi 👋 

This repo is fantastic to see and so far its been brilliant. 

So far I followed the pre requisite steps to get setup: 
```
make vendor
make
kubectl apply -f manifests
```
Here's a snapshot of my cluster after applying all of the manifests
<img width="1107" alt="Screenshot 2020-03-17 at 20 18 08" src="https://user-images.githubusercontent.com/15033304/76898075-e1ba1400-688c-11ea-8815-f8791d1deb9f.png">

Everything looks great! 

Unfortunately though I am unable to access the ingress gateways as suggested alertmanager.<IP>.nip.io, prometheus.<IP>.nip.io, grafana.<IP>.nip.io. To get around this for now I thought I'd check out prometheus first via port-forward.

To do that I run: `kubectl -n monitoring port-forward prometheus-k8s-0 9090:9090` 

When I check prometheus targets I see lots of similar errors to this one:
<img width="1893" alt="Screenshot 2020-03-17 at 20 25 28" src="https://user-images.githubusercontent.com/15033304/76898444-a23ff780-688d-11ea-911b-ab343c40f58c.png"> 

The majority of the errors are:
* context deadline exceeded
* IO Timeout

Details of my hardware:
```
NAME               STATUS   ROLES    AGE   VERSION
worker-rpi-3Bplus   Ready    <none>   22h   v1.17.3+k3s1
worker-rpi-2       Ready    <none>   22h   v1.17.3+k3s1
master             Ready    master   22h   v1.17.3+k3s1
worker-rpi-4       Ready    <none>   22h   v1.17.3+k3s1
```

master = Raspberry Pi 4 /w 4GB RAM
worker-rpi-4 = Rapsberry Pi 4 /w 4GB RAM
worker-rpi-2 = Raspberry Pi 2 B+
worker-rpi-3Bplus = Raspberry Pi 3 B +

All running docker version: 19.03.8 Client & Server. 

If there's any further detail you need me to supply, I'll dig that out for you.

There also may be a gap in my knowledge here / potential further PEBCAC. Any advise / help would be much appreciated!

**Comment by james-millner** [comment]
Oh and apologies I left out my network policy that exists in the default namespace. If that is required too for any diagnostics

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: default
spec:
  podSelector: {}
  policyTypes:
  - Ingress
  - Egress
```

**Comment by james-millner** [comment]
After reading some of the previous issues. I thought I'd share all the exporters that haven't registered with prometheus:

- monitoring/alertmanager/0 (0/1 up)
- monitoring/coredns/0 (0/2 up)
- monitoring/grafana/0 (0/1 up) 
- monitoring/kube-state-metrics/0 (0/1 up)
- monitoring/kube-state-metrics/1 (0/1 up)
- monitoring/kubelet/0 (0/4 up)
- monitoring/kubelet/1 (0/4 up)
- monitoring/node-exporter/0 (1/4 up)  (Master up, but none of the workers)
- monitoring/prometheus-operator/0 (0/1 up)
- monitoring/traefik/0 (0/1 up)

**Comment by carlosedp** [comment]
Hi @james-millner, most problems I've seen related to these messages is where Prometheus can't or times-out scraping the endpoints.

I have two scenarios here, one a VirtualBox VM with two network interfaces and the first is not the exposed one. On this VM, I have similar problems as you and I even commented on a similar problem with K3s (https://github.com/rancher/k3s/issues/1216).

On the second scenario, I took an ARM64 board and deployed latest K3s (https://github.com/rancher/k3s/issues/1216) and my updated stack with latest vendor libs and some adjustments. 

Everything went well:

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/20382/76971570-1f11b680-690c-11ea-85ed-6dc373ad3417.png">

<img width="1174" alt="image" src="https://user-images.githubusercontent.com/20382/76971612-2b960f00-690c-11ea-878b-663265a8ce67.png">

<img width="1499" alt="image" src="https://user-images.githubusercontent.com/20382/76971655-38b2fe00-690c-11ea-8c62-50c86e073236.png">

Take a look if this multiple NIC scenario is similar or try with just one node freshly installed to begin with debugging.

**Comment by james-millner** [comment]
Hi Carlos, 

Sorry for the delay in getting back to you. 

Well I've had some success! It appears to be perhaps a dodgy node within my cluster but I cannot fully say - this is what it appears to be 🤔 .  

I uninstalled K3s completely and only re-installed it on the Raspberry Pi 4's & Raspberry Pi 3 B+. 

I also installed K3s with the docker flag overriding containerd and WHAM, it all works perfectly 👌. Thank you so much for this fantastic repository & contribution!

![Screenshot 2020-03-20 at 14 17 10](https://user-images.githubusercontent.com/15033304/77172305-a77f8b00-6ab5-11ea-87de-bebc30d3e88c.png)
